### ISSUE TITLE
Create an initial web components design

### ISSUE THREAD

**Issue by nicojs** [stryker-mutator/mutation-testing-elements#2]
We should try to come up with a software design for the web components that we want to create.

On a high level, I think we will have 3 components:

| Name | Responsibility | Example |
|--|--|--|
| `mutation-testing-file-report` | Display a single file result | `<mutation-testing-file-report>` |
| `mutation-testing-directory-report` | Display a directory result. Deep link to files using the [window.location.hash](https://developer.mozilla.org/en-US/docs/Web/API/Location). | `<mutation-testing-directory-report>` |
| `mutation-testing-report-app` | Display the entire report. Load current location using the location hash. | `<mutation-testing-report-app src="/url/to/report.json">` |

There are other smaller components, like breadcrumb, mutation totals table, etc.

I think we shouldn't rely on any library/framework for these components. Let's try it with plain web components.

**Comment by nicojs** [comment]
Friends, I've started implementation for this feature on the [feat-html-elements branch](https://github.com/stryker-mutator/mutation-testing-elements/tree/feat-html-elements):

![image](https://user-images.githubusercontent.com/1828233/52976419-b931c300-33c9-11e9-9a25-eb0fc6feb667.png)

@stryker-mutator/stryker4s  @stryker-mutator/stryker-net 
Could you try it out on Stryker.NET and Stryker4s and give feedback? I would also love for you to provide an actual `mutation-report.json` that I can use for testing. I'm also planning automated tests with it. A moderate or small project is good enough for now.

I released it under `0.0.3` so we can test with it. You can download it here:

https://unpkg.com/mutation-testing-elements@0.0.3/dist/mutation-test-elements.js

In order to test it:

1. Download the js file, put it next to your `mutation-report.json`.
2. Create an `index.html` file in this format:
    ```html
    <!DOCTYPE html>
    <head>
        <script src="/mutation-test-elements.js"></script>
    </head>
    <body>
        <a href="/">Back</a></li>
        <mutation-test-report-app src="mutation-report.json"></mutation-test-report-app>
    </body>
    </html>
    ```
   (yep, it's awesomely simple like that 😘).
 
That's it. One caveat, Chrome gives this error: `Fetch API cannot load file:///C:/z/github/stryker-mutator/mutation-testing-elements/testResources/simple-example/mutation-report.json. URL scheme must be "http" or "https" for CORS request`. There are 3 sollutions:

1. Force users to disable the security feature: https://stackoverflow.com/questions/8449716/cross-origin-requests-are-only-supported-for-http-but-its-not-cross-domain/14490982#14490982 or use firefox 🙄
2. Allow data to be loaded with [JSONP](https://en.wikipedia.org/wiki/JSONP) 🤐
3. Allow data to be loaded directly from an attribute: `<mutation-test-report-app report-data="{ &quot;name&quot;: &quot;src&quot; /*...*/}"></mutation-test-report-app>`

For now, we can use 1 or 2. I think we should implement the 3rd option for an official release.

*Note: This is not a problem for the dashboard usecase, as we can just load the json from the same domain over https.*

I've updated the json format a bit (not yet updated the schema, first want to see if we all agree), you can see the updates at #1. I'll post a new message there detailing what changed and why.

**Comment by nicojs** [comment]
From @Mobrockers:

This report does not seem to work:
[mutation-report.zip](https://github.com/stryker-mutator/mutation-testing-elements/files/2879668/mutation-report.zip)

**Comment by nicojs** [comment]
@Mobrockers I've found the issue. Deep linking wasn't supported yet. I've published 0.0.4, can you try again?

**Comment by Mobrockers** [comment]
Done, it works!

**Comment by Mobrockers** [comment]
The syntax highlighting on survived mutations seems to sometimes randomly colour a single letter in the expanded mutation.

![image](https://user-images.githubusercontent.com/3595932/53042120-daf08000-3485-11e9-903b-d2df34e21c47.png)
![image](https://user-images.githubusercontent.com/3595932/53042171-f2c80400-3485-11e9-90a9-b02a11310574.png)

**Comment by nicojs** [comment]
I've just released 0.0.5, with a first implementation for the new flat JSON syntax of #1. Happy testing, I will do the same 👍

**Comment by hugo-vrijswijk** [comment]
Got the json working today for Stryker4s. Here is a small example of our first implementation: https://gist.github.com/hugo-vrijswijk/bbb696635f0f89ede1b315e1564a4594

Note: there seems to be a bug with the highlighting if the mutation is the last part of a line, see `MutantRunResultMapper.scala` in the example. Everything below the mutation is also red (and crossed out):
![image](https://user-images.githubusercontent.com/10114577/53183197-28422e00-35fb-11e9-9812-97f3155a2a5d.png)

**Comment by nicojs** [comment]
> Note: there seems to be a bug with the highlighting if the mutation is the last part of a line, see `MutantRunResultMapper.scala` in the example. Everything below the mutation is also red (and crossed out):

I fixed this issue. The problem was that the mutant-component was that mutant 179 ended at `"line" : 13, "column" : 34`. However, per my calculation, that line only has 33 columns. Fixed it by closing mutants before the next character of the new line if this occurs. I released  `0.0.6` with the fix.

> Also, it seems to cut out to the most relative path. So from `core/src/main/scala/stryker4s/report/mapper/MutantRunResultMapper.scala` it just makes `/report/mapper/MutantRunResultMapper.scala` which looks a bit weird

Hmm. I understand the confusion. You want me to remove the logic of finding the biggest common denominator? I'm not sure what the best thing is here.

**Comment by hugo-vrijswijk** [comment]
> I released 0.0.6 with the fix.

Huzzah! 🎉

> You want me to remove the logic of finding the biggest common denominator?

Not sure. I'd be interested in what the others think. To me, a view similar to GitHub would make sense: show all folders, skip through empty ones, and show the score of that folder

**Comment by Wmaarts** [comment]
> a view similar to GitHub would make sense

Second this. But perhaps an easier / quicker improvement would be to show the common denominator somewhere at the top. Maybe where it says 'All files': first column first row of the table. :) 
Also, removing the `/` at the start of each file name already looks a lot better imo.

**Comment by nicojs** [comment]
Could you create a separate issue for this with examples of how you want the tables to look like? We can discuss that part further there. I first want to create a 1.0 version of the schema en elements. We can improve iteratively after that.

**Comment by nicojs** [comment]
The latest version (`0.0.7`) looks like this:

![image](https://user-images.githubusercontent.com/1828233/53473073-bf7b0b80-3a69-11e9-8bdc-f5082c106dc6.png)

This version also supports _embedded report json data_. It works by setting the `report` property directly:

```html
    <mutation-test-report-app title-postfix="Stryker report"></mutation-test-report-app>
    <script src="/mutation-test-elements.js"></script>
    <script>
        document.querySelector('mutation-test-report-app').report = {
          "schemaVersion" : "1",
          "thresholds" : {
            "high" : 80,
            "low" : 60
          },
          "files" : {
                /* ... */
          }
      };
    </script>
```

**Note:** Setting a property directly on the custom element is actually the web components way of binding to a property. A neat little detail 😎 .

As you can also see in the example, I've added a `title-postfix` attribute. You can use that to set the title postfix. I.e. "Stryker report" or "Stryker4s" or something.

If everyone is happy, I'm planning on releasing v0.1 (after I've written some more UI, integration and unit tests).

**Comment by Mobrockers** [comment]
@nicojs Here is our current v0.0.7 stryker-net html report :)

https://gist.github.com/Mobrockers/7a48aa9d3ca091b5551d02b084c0bac9

**Comment by nicojs** [comment]
This is realized with #10. Released in v`1.0.0`.
